doc/man/doveadm-who.1.in
.\" Copyright (c) 2010-2018 Dovecot authors, see the included COPYING file
.TH DOVEADM\-WHO 1 "2010-07-12" "Dovecot v2.3" "Dovecot"
.SH NAME
doveadm\-who \- Show who is logged in to the Dovecot server
.\"------------------------------------------------------------------------
.SH SYNOPSIS
.BR doveadm " [" \-Dv "] [" \-f
.IR formatter ]
.BR who\  [ \-1 ]
.RB [ -f
.IR passdb_field ]
[\fB\-a\fP \fIanvil_socket_path\fP]
.RI [ user_mask ]
[\fIip\fP[\fB/\fP\fIbits\fP]]
.\"------------------------------------------------------------------------
.SH DESCRIPTION
The
.B who
command is used to show which users from which hosts are currently
connected to which service.
.\"------------------------------------------------------------------------
@INCLUDE:global-options-formatter@
.\" --- command specific options --- "/.
.PP
This command uses by default the output formatter
.BR table .
.PP
Command specific
.IR options :
.\"-------------------------------------
.TP
.B \-1
Print one line per user and connection.
Otherwise the connections are grouped by the username.
.\"-------------------------------------
.TP
.BI \-a\  anvil_socket_path
This option is used to specify an alternative socket.
The option\(aqs argument is either an absolute path to a local UNIX domain
socket, or a hostname and port
.RI ( hostname : port ),
in order to connect a remote host via a TCP socket.
.sp
By default
.BR doveadm (1)
will use the socket
.IR @rundir@/anvil .
The socket may be located in another directory, when the default
.I base_dir
setting was overridden in
.IR @pkgsysconfdir@/dovecot.conf .
.\"-------------------------------------
.TP
.BI \-f\  passdb_field
Alternative username field to use for kicking, as returned by passdb.
Only the passdb fields beginning with the
.I user_
prefix are tracked.
.\"------------------------------------------------------------------------
.SH ARGUMENTS
.TP
\fIip\fP[\fB/\fP\fIbits\fP]
Specify an
.I ip
address or network range, in CIDR notation, to reduce the result to
matching connections.
.\"-------------------------------------
.TP
.I user_mask
List only users whose login name matches the
.IR user_mask ,
or the altenative username (user_* field) if the
.B -f
parameter is used.
It\(aqs also possible to use wildcards in the
.I user
name.
.\"------------------------------------------------------------------------
.SH EXAMPLE
Show authenticated sessions, filtered by the client\(aqs IP address.
.sp
.nf
.B doveadm who ::1
username                       # proto (pids)        (ips)
jane                           2 imap  (30155 30412) (::1)
.B doveadm who 192.0.2.0/24
username                        # proto (pids)  (ips)
john@example.com                1 imap  (30257) (192.0.2.34)
.fi
.PP
Show authenticated sessions, filtered by username.
.sp
.nf
.B doveadm who pvo
username         # proto (pids)                    (ips)
pvo              1 sieve (30343)                   (fd95:4eed:38ba::25)
pvo              4 imap  (25693 25686 25679 25669) (fd95:4eed:38ba::25)
.B doveadm who ja*
username                    # proto (pids)        (ips)
james                       1 imap  (30091)       (127.0.0.1)
jane                        2 imap  (30155 30412) (::1)
.fi
.\"------------------------------------------------------------------------
@INCLUDE:reporting-bugs@
.\"------------------------------------------------------------------------
.SH SEE ALSO
.BR doveadm (1),
.BR doveadm\-kick (1)

NEWS
v2.3.18 2022-02-03  Aki Tuomi <aki.tuomi@open-xchange.com>

	* Removed mail_cache_lookup_finished event. This event wasn't especially
	  useful, but it increased CPU usage significantly.
	* fts: Don't index inline base64 encoded content in FTS indexes using
	  the generic tokenizer. This reduces the FTS index sizes by removing
	  input that is very unlikely to be searched for. See
	  https://doc.dovecot.org/configuration_manual/fts/tokenization for
	  details on how base64 is detected. Only applies when using libfts.
	* lmtp: Session IDs are now preserved through proxied connections, so
	  LMTP sessions can be tracked. This slightly changes the LMTP session
	  ID format by appending ":Tn" (transaction), ":Pn" (proxy connection)
	  and ":Rn" (recipient) counters after the session ID prefix.
	+ Events now have "reason_code" field, which can provide a list of
	  reasons why the event is happening. See
	  https://doc.dovecot.org/admin_manual/event_reasons/
	+ New events are added. See https://doc.dovecot.org/admin_manual/list_of_events/
	+ fts: Added fts_header_excludes and fts_header_includes settings to
	  specify which headers to index. See
	  https://doc.dovecot.org/settings/plugin/fts-plugin#plugin-fts-setting-fts-header-excludes
	  for configuration details.
	+ fts: Initialize the textcat language detection library only once per
	  process. This can reduce CPU usage if fts_languages setting has multiple
	  languages listed and service indexer-worker { service_count } isn't 1.
	  Only applies when using libfts.
	+ lib-storage: Reduced CPU usage significantly for some operations that
	  accessed lots of emails (e.g. fetching all flags in a folder, SORT, ...)
	+ lib: DOVECOT_PREREQ() - Add micro version which enables compiling
	  external plugins against different versions of Dovecot.
	+ lmtp: Added new lmtp_verbose_replies setting that makes errors sent to
	  the LMTP client much more verbose with details about why exactly
	  backend proxy connections or commands are failing.
	+ submission: Support implicit SASL EXTERNAL with
	  submission_client_workarounds=implicit-auth-external. This allows
	  automatically logging in when SSL client certificate is present.
	- *-login: Statistics were disabled if stats process connection was lost.
	- auth: Authentication master user login fails with SCRAM-* SASL mechanisms.
	- auth: With auth_cache_verify_password_with_worker=yes, passdb extra
	  fields in the auth cache got lost.
	- doveadm: Fixed crash if zlib_save_level setting was specified,
	  but zlib_save was unset. v2.3.15 regression.
	- doveadm: Proxying can panic when flushing print output. v2.3.17
	  regression. Fixes:
	  Panic: file ioloop.c: line 865 (io_loop_destroy): assertion failed:
	  (ioloop == current_ioloop)
	- doveadm: stats add --group-by parameter didn't work.
	- fts: Using email-address fts tokenizer could result in excessive memory
	  usage with garbage email input. This could cause the indexer-worker
	  processes to fail due to reaching the VSZ memory size limit.
	  Only applies when using libfts.
	- imap: A SEARCH command timing out while fts returns indexes may timeout
	  returning "NO [SERVERBUG]", while it should return "NO [INUSE]" instead.
	- imap: LIST-EXTENDED doesn't return STATUS for all folders. Sending
	  LIST .. RETURN (SUBSCRIBED STATUS (...)) did not return STATUS for
	  folders that are not subscribed when they have a child folder that is
	  subscribed as mandated by IMAP RFCs.
	- imapc: Mailbox vsize calculation crashed with
	  Panic: file index-mailbox-size.c: line 344 (index_mailbox_vsize_hdr_add_missing):
	  assertion failed: (mails_left > 0)
	- indexer: If indexer-worker crashes, the request it was processing gets
	  stuck in the indexer process. This stops indexing for the folder until
	  indexer process is restarted. v2.3.14 regression.
	- indexer: Process was slowly leaking memory for each indexing request.
	- lib-event: Unnamed events were wrongly filtered out for event/metric
	  filters like "event=abc OR something_independent_of_event_name".
	- lib-index: 64-bit big endian CPUs handle last_used field in
	  dovecot.index.cache wrong.
	- lib-ssl-iostream: Fix buggy OpenSSL error handling without assert-crashing.
	  If there is no error available, log it as an error instead of crashing.
	  The previous fix for this in v2.3.11 was incomplete. Fixes
	  Panic: file istream-openssl.c: line 51 (i_stream_ssl_read_real):
	  assertion failed: (errno != 0)
	- lmtp: Out-of-memory issues can happen when proxying large messages to
	  LMTP backend servers that accept the message data too slow.
	- master: HAProxy header parsing has read buffer overflow if provided
	  header size is invalid. This happens only if inet_listener
	  { haproxy=yes } is configured and only if the remote IP address is in
	  haproxy_trusted_networks.
	- old_stats: Plugin kept increasing memory usage, which became
	  noticeable with long-running imap sessions.
	- stats: Dynamically adding same metric multiple times causes multiple stats.
	- submission-login: Authentication does not accept OAUTH2 token (or
	  other very long credentials) because it considers the line to be too long.
	- submission-login: Process can crash if HELO is pipelined with an
	  invalid domain.
	- submission-proxy: Don't use SASL-IR if it would make the AUTH command
	  line longer than 512 bytes.
	- submission: Service would crash if relay server authentication failed.
	- virtual: FTS search in a virtual folder could crash if there are
	  duplicate mailbox GUIDs. This mainly happened when user had both INBOX
	  and INBOX/INBOX folders and the namespace prefix was INBOX/. Fixes
	  Panic: file hash.c: line 252 (hash_table_insert_node):
	  assertion failed: (opcode == HASH_TABLE_OP_UPDATE)
	- virtual: If mailbox opening fails, the backend mailbox is leaked and
	  process crashes when client disconnects. Fixes
	  Panic: file mail-user.c: line 232 (mail_user_deinit):
	  assertion failed: ((*user)->refcount == 1)

src/doveadm/doveadm-who.c
/* Copyright (c) 2009-2018 Dovecot authors, see the included COPYING file */

#include "lib.h"
#include "array.h"
#include "net.h"
#include "istream.h"
#include "wildcard-match.h"
#include "hash.h"
#include "str.h"
#include "strescape.h"
#include "doveadm.h"
#include "doveadm-print.h"
#include "doveadm-who.h"

#include <stdio.h>
#include <unistd.h>

struct who_user {
	const char *username;
	const char *service;
	ARRAY(struct ip_addr) ips;
	ARRAY(pid_t) pids;
	unsigned int connection_count;
};

struct doveadm_who_iter {
	struct istream *input;
	pool_t pool, line_pool;
	bool failed;
};

static void who_user_ip(const struct who_user *user, struct ip_addr *ip_r)
{
	if (array_count(&user->ips) == 0)
		i_zero(ip_r);
	else {
		const struct ip_addr *ip = array_front(&user->ips);
		*ip_r = *ip;
	}
}

static unsigned int who_user_hash(const struct who_user *user)
{
	struct ip_addr ip;
	unsigned int hash = str_hash(user->service);

	if (user->username[0] != '\0')
		hash += str_hash(user->username);
	else {
		who_user_ip(user, &ip);
		hash += net_ip_hash(&ip);
	}
	return hash;
}

static int who_user_cmp(const struct who_user *user1,
			const struct who_user *user2)
{
	if (strcmp(user1->username, user2->username) != 0)
		return 1;
	if (strcmp(user1->service, user2->service) != 0)
		return 1;

	if (user1->username[0] == '\0') {
		/* tracking only IP addresses, not usernames */
		struct ip_addr ip1, ip2;

		who_user_ip(user1, &ip1);
		who_user_ip(user2, &ip2);
		return net_ip_cmp(&ip1, &ip2);
	}
	return 0;
}

static bool
who_user_has_ip(const struct who_user *user, const struct ip_addr *ip)
{
	const struct ip_addr *ex_ip;

	array_foreach(&user->ips, ex_ip) {
		if (net_ip_compare(ex_ip, ip))
			return TRUE;
	}
	return FALSE;
}

static int who_parse_line(struct doveadm_who_iter *iter,
			  const char *line, struct who_line *line_r)
{
	const char *const *args = t_strsplit_tabescaped(line);
	i_zero(line_r);

	/* <pid> <username> <service> <ip> <conn-guid> <dest-ip> */
	if (str_array_length(args) < 6)
		return -1;

	p_clear(iter->line_pool);
	if (str_to_pid(args[0], &line_r->pid) < 0)
		return -1;
	line_r->username = p_strdup(iter->line_pool, args[1]);
	line_r->service = p_strdup(iter->line_pool, args[2]);
	if (args[3][0] != '\0') {
		if (net_addr2ip(args[3], &line_r->ip) < 0)
			return -1;
	}
	if (guid_128_from_string(args[4], line_r->conn_guid) < 0)
		return -1;
	if (args[5][0] != '\0') {
		if (net_addr2ip(args[5], &line_r->dest_ip) < 0)
			return -1;
	}
	return 0;
}

static bool who_user_has_pid(struct who_user *user, pid_t pid)
{
	pid_t ex_pid;

	array_foreach_elem(&user->pids, ex_pid) {
		if (ex_pid == pid)
			return TRUE;
	}
	return FALSE;
}

static void who_aggregate_line(struct who_context *ctx,
			       const struct who_line *line)
{
	struct who_user *user, lookup_user;

	lookup_user.username = line->username;
	lookup_user.service = line->service;

	user = hash_table_lookup(ctx->users, &lookup_user);
	if (user == NULL) {
		user = p_new(ctx->pool, struct who_user, 1);
		user->username = p_strdup(ctx->pool, line->username);
		user->service = p_strdup(ctx->pool, line->service);
		p_array_init(&user->ips, ctx->pool, 3);
		p_array_init(&user->pids, ctx->pool, 8);
		hash_table_insert(ctx->users, user, user);
	}
	user->connection_count++;

	if (line->ip.family != 0 && !who_user_has_ip(user, &line->ip))
		array_push_back(&user->ips, &line->ip);

	if (!who_user_has_pid(user, line->pid))
		array_push_back(&user->pids, &line->pid);
}

int who_parse_args(struct who_context *ctx, const char *const *masks)
{
	struct ip_addr net_ip;
	unsigned int i, net_bits;

	for (i = 0; masks[i] != NULL; i++) {
		if (net_parse_range(masks[i], &net_ip, &net_bits) == 0) {
			if (ctx->filter.net_bits != 0) {
				i_error("Multiple network masks not supported");
				doveadm_exit_code = EX_USAGE;
				return -1;
			}
			ctx->filter.net_ip = net_ip;
			ctx->filter.net_bits = net_bits;
		} else {
			if (ctx->filter.username != NULL) {
				i_error("Multiple username masks not supported");
				doveadm_exit_code = EX_USAGE;
				return -1;
			}
			ctx->filter.username = masks[i];
		}
	}
	return 0;
}

struct doveadm_who_iter *doveadm_who_iter_init(const char *anvil_path)
{
#define ANVIL_HANDSHAKE "VERSION\tanvil\t2\t0\n"
#define ANVIL_CMD ANVIL_HANDSHAKE"CONNECT-DUMP\n"
	struct doveadm_who_iter *iter;
	const char *line;
	int fd;
	pool_t pool;

	pool = pool_alloconly_create("doveadm who iter", 256);
	iter = p_new(pool, struct doveadm_who_iter, 1);
	iter->pool = pool;
	iter->line_pool = pool_alloconly_create("doveadm who line", 256);

	fd = doveadm_connect(anvil_path);
	net_set_nonblock(fd, FALSE);
	if (write(fd, ANVIL_CMD, strlen(ANVIL_CMD)) < 0) {
		i_error("write(%s) failed: %m", anvil_path);
		iter->failed = TRUE;
		return iter;
	}

	iter->input = i_stream_create_fd_autoclose(&fd, SIZE_MAX);
	i_stream_set_name(iter->input, anvil_path);
	if ((line = i_stream_read_next_line(iter->input)) == NULL) {
		i_error("anvil didn't send header line");
		iter->failed = TRUE;
	}
	return iter;
}

bool doveadm_who_iter_next(struct doveadm_who_iter *iter,
			   struct who_line *who_line_r)
{
	const char *line;
	int ret;

	if (iter->failed)
		return FALSE;

	if ((line = i_stream_read_next_line(iter->input)) != NULL) {
		if (*line == '\0')
			return FALSE;
		T_BEGIN {
			ret = who_parse_line(iter, line, who_line_r);
		} T_END;
		if (ret < 0) {
			i_error("Invalid input: %s", line);
			iter->failed = TRUE;
			return FALSE;
		}
		return TRUE;
	}
	if (iter->input->stream_errno != 0) {
		i_error("read(%s) failed: %s", i_stream_get_name(iter->input),
			i_stream_get_error(iter->input));
	} else {
		i_error("read(%s) failed: Unexpected EOF",
			i_stream_get_name(iter->input));
	}
	iter->failed = TRUE;
	return FALSE;
}

int doveadm_who_iter_deinit(struct doveadm_who_iter **_iter)
{
	struct doveadm_who_iter *iter = *_iter;
	bool failed = iter->failed;

	*_iter = NULL;
	i_stream_destroy(&iter->input);
	pool_unref(&iter->line_pool);
	pool_unref(&iter->pool);
	return failed ? -1 : 0;
}

static bool who_user_filter_match(const struct who_user *user,
				  const struct who_filter *filter)
{
	if (filter->username != NULL) {
		if (!wildcard_match_icase(user->username, filter->username))
			return FALSE;
	}
	if (filter->net_bits > 0) {
		const struct ip_addr *ip;
		bool ret = FALSE;

		array_foreach(&user->ips, ip) {
			if (net_is_in_network(ip, &filter->net_ip,
					      filter->net_bits)) {
				ret = TRUE;
				break;
			}
		}
		if (!ret)
			return FALSE;
	}
	return TRUE;
}

static void who_print_user(const struct who_user *user)
{
	const struct ip_addr *ip;
	pid_t pid;
	string_t *str = t_str_new(256);

	doveadm_print(user->username);
	doveadm_print(dec2str(user->connection_count));
	doveadm_print(user->service);

	str_append_c(str, '(');
	array_foreach_elem(&user->pids, pid)
		str_printfa(str, "%ld ", (long)pid);
	if (str_len(str) > 1)
		str_truncate(str, str_len(str)-1);
	str_append_c(str, ')');
	doveadm_print(str_c(str));

	str_truncate(str, 0);
	str_append_c(str, '(');
	array_foreach(&user->ips, ip)
		str_printfa(str, "%s ", net_ip2addr(ip));
	if (str_len(str) > 1)
		str_truncate(str, str_len(str)-1);
	str_append_c(str, ')');
	doveadm_print(str_c(str));

	doveadm_print_flush();
}

static void who_print(struct who_context *ctx)
{
	struct hash_iterate_context *iter;
	struct who_user *user;

	doveadm_print_header("username", "username", 0);
	doveadm_print_header("connections", "#",
			     DOVEADM_PRINT_HEADER_FLAG_RIGHT_JUSTIFY);
	doveadm_print_header("service", "proto", 0);
	doveadm_print_header("pids", "(pids)", 0);
	doveadm_print_header("ips", "(ips)", 0);

	iter = hash_table_iterate_init(ctx->users);
	while (hash_table_iterate(iter, ctx->users, &user, &user)) {
		if (who_user_filter_match(user, &ctx->filter)) T_BEGIN {
			who_print_user(user);
		} T_END;
	}
	hash_table_iterate_deinit(&iter);
}

bool who_line_filter_match(const struct who_line *line,
			   const struct who_filter *filter)
{
	if (filter->username != NULL) {
		if (!wildcard_match_icase(line->username, filter->username))
			return FALSE;
	}
	if (filter->net_bits > 0) {
		if (!net_is_in_network(&line->ip, &filter->net_ip,
				       filter->net_bits))
			return FALSE;
	}
	return TRUE;
}

static void who_print_line(struct who_context *ctx,
			   const struct who_line *line)
{
	if (!who_line_filter_match(line, &ctx->filter))
		return;

	doveadm_print(line->username);
	doveadm_print(line->service);
	doveadm_print(dec2str(line->pid));
	doveadm_print(net_ip2addr(&line->ip));
	doveadm_print(net_ip2addr(&line->dest_ip));
}

static void cmd_who(struct doveadm_cmd_context *cctx)
{
	const char *const *masks;
	struct who_context ctx;
	bool separate_connections = FALSE;

	i_zero(&ctx);
	if (!doveadm_cmd_param_str(cctx, "socket-path", &(ctx.anvil_path)))
		ctx.anvil_path = t_strconcat(doveadm_settings->base_dir, "/anvil", NULL);
	(void)doveadm_cmd_param_bool(cctx, "separate-connections", &separate_connections);

	ctx.pool = pool_alloconly_create("who users", 10240);
	hash_table_create(&ctx.users, ctx.pool, 0, who_user_hash, who_user_cmp);

	if (doveadm_cmd_param_array(cctx, "mask", &masks)) {
		if (who_parse_args(&ctx, masks) != 0) {
			hash_table_destroy(&ctx.users);
			pool_unref(&ctx.pool);
			return;
		}
	}

	doveadm_print_init(DOVEADM_PRINT_TYPE_TABLE);
	struct doveadm_who_iter *iter = doveadm_who_iter_init(ctx.anvil_path);
	struct who_line who_line;
	if (!separate_connections) {
		while (doveadm_who_iter_next(iter, &who_line))
			who_aggregate_line(&ctx, &who_line);
		who_print(&ctx);
	} else {
		doveadm_print_header("username", "username",
				     DOVEADM_PRINT_HEADER_FLAG_EXPAND);
		doveadm_print_header("service", "proto", 0);
		doveadm_print_header_simple("pid");
		doveadm_print_header_simple("ip");
		doveadm_print_header_simple("dest_ip");
		while (doveadm_who_iter_next(iter, &who_line))
			who_print_line(&ctx, &who_line);
	}
	if (doveadm_who_iter_deinit(&iter) < 0)
		doveadm_exit_code = EX_TEMPFAIL;

	hash_table_destroy(&ctx.users);
	pool_unref(&ctx.pool);
}

struct doveadm_cmd_ver2 doveadm_cmd_who_ver2 = {
	.name = "who",
	.cmd = cmd_who,
	.usage = "[-a <anvil socket path>] [-1] [<user mask>] [<ip/bits>]",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('a',"socket-path", CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAM('1',"separate-connections", CMD_PARAM_BOOL, 0)
DOVEADM_CMD_PARAM('\0',"mask", CMD_PARAM_ARRAY, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
};
